model: Add BMRetriever

whybe-choi opened the pull request:
Close #3194

- [x] I have filled out the ModelMeta object to the extent possible
- [x] I have ensured that my model can be loaded using
  - [x] `mteb.get_model(model_name, revision)` and
  - [x] `mteb.get_model_meta(model_name, revision)`
- [x] I have tested the implementation works on a representative set of tasks.
- [x] The model is public, i.e. is available either as an API or the wieght are publicly avaiable to download

whybe-choi commented on mteb/models/bmretriever_models.py:
```diff
@@ -0,0 +1,256 @@
+    ):
+        super().__init__(**kwargs)
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        self.model = AutoModel.from_pretrained(
```
Thanks a lot for the helpful reference!

I'm not the owner of the model repository, so I don't have permission to upload or modify files directly on the Hugging Face Hub 🥲 I'm planning to submit a PR for the files you mentioned, but I wanted to ask if it's acceptable to contribute the code in the following format:

```python
class BMRetrieverWrapper(SentenceTransformerWrapper):
    """Wrapper for the BMRetriever models built on top of SentenceTransformerWrapper."""

    def __init__(
        self,
        model_name_or_path: str,
        instruction_template: Callable[[str, PromptType | None], str],
        *,
        max_seq_length: int = 512,
        torch_dtype: torch.dtype = torch.float32,
        do_lower_case: bool = False,
        add_eos_token: bool = True,
        **kwargs: Any,
    ) -> None:
        self.instruction_template = instruction_template
        self.add_eos_token = add_eos_token

        transformer = Transformer(
            model_name_or_path,
            max_seq_length=max_seq_length,
            do_lower_case=do_lower_case,
            model_kwargs={
                "torch_dtype": torch_dtype,
            },
        )
        pooling = Pooling(transformer.get_word_embedding_dimension(), pooling_mode="lasttoken")
        sentence_model = SentenceTransformer(modules=[transformer, pooling])

        if sentence_model.tokenizer.pad_token is None:
            sentence_model.tokenizer.pad_token = sentence_model.tokenizer.eos_token
            sentence_model.tokenizer.padding_side = "left"
        sentence_model.max_seq_length = max_seq_length

        super().__init__(model=sentence_model, **kwargs)
```

Samoed commented on mteb/models/bmretriever_models.py:
```diff
@@ -0,0 +1,252 @@
+    max_tokens=2048,
+    license="mit",
+    reference="https://huggingface.co/BMRetriever/BMRetriever-410M",
+    similarity_fn_name="dot",
```
I think model actually use `cosine` function
```suggestion
    similarity_fn_name="cosine",
```

whybe-choi commented:
Great work on implementing the BMRetriever model! The changes look good overall. I particularly like how you've addressed the reviewer feedback and migrated to using InstructSentenceTransformerWrapper as suggested. The implementation follows the project's conventions well.

whybe-choi commented:
@Samoed 
Could you please review this PR again? I have updated the implementation according to your feedback. Thank you!

Samoed reviewed (approved):
Great!

KennethEnevoldsen commented on mteb/models/bmretriever_models.py:
```diff
@@ -0,0 +1,190 @@
+            self.model.tokenizer.padding_side = padding_side
+
+
+BMRETRIEVER_TRAINING_DATA = {
+    "MedRAG/pubmed": ["train"],
+    "mteb/raw_arxiv": ["train"],
+    "medalpa/medical_meadow_cord19": ["train"],
+    "MedRAG/textbooks": ["train"],
+    "MedRAG/statpearls": ["train"],
+    "KushT/LitCovid_BioCreative": ["train"],
+    "S2ORC": ["train"],
+    "MSMARCO": ["train"],
+    "BMRetriever/biomed_retrieval_dataset": ["train"],
+}
```
These have to be MTEB datasets, where the key is the task name. I would just comment out the rest:
```suggestion
BMRETRIEVER_TRAINING_DATA = {
    "MedRAG/pubmed": ["train"],
    "mteb/raw_arxiv": ["train"],
    "medalpa/medical_meadow_cord19": ["train"],
    "MedRAG/textbooks": ["train"],
    "MedRAG/statpearls": ["train"],
    "KushT/LitCovid_BioCreative": ["train"],
    "S2ORC": ["train"],
    "MSMARCO": ["train"],
    # non-mteb
    # "BMRetriever/biomed_retrieval_dataset": ["train"],
}
```

whybe-choi replied:
So, do you mean I should change it like this, to only keep MSMARCO?

```python
BMRETRIEVER_TRAINING_DATA = {
    "MSMARCO": ["train"],
    # non-mteb
    # "mteb/raw_arxiv": ["train"],
    # "MedRAG/pubmed": ["train"],
    # "medalpa/medical_meadow_cord19": ["train"],
    # "MedRAG/textbooks": ["train"],
    # "MedRAG/statpearls": ["train"],
    # "KushT/LitCovid_BioCreative": ["train"],
    # "S2ORC": ["train"],
    # "BMRetriever/biomed_retrieval_dataset": ["train"],
}
```

KennethEnevoldsen reviewed (approved):
besides the comment then I think this is good to merge
